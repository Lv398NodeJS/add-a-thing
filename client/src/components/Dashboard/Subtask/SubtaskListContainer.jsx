import React, { Component } from 'react';
import { Row, Col, Container } from 'react-bootstrap';
import { connect } from 'react-redux';
import SubtaskProgressBar from './SubtaskProgressBar/SubtaskProgressBar';
import SubtaskList from './SubtaskList/SubtaskList';
import SubtaskFilterContainer from './SubtaskFilter/SubtaskFilterContainer';
import SubtaskAdd from './SubtaskAdd/SubtaskAdd';

export class SubtaskListContainer extends Component {
  render() {
    const {
      taskId,
      taskStatus,
      subtaskList = {},
      currentFilter,
    } = this.props;
    return (
      <Container fluid className="my-0 mx-0">
        <Row>
          <Col className="col-4">
            <h5>Subtasks:</h5>
          </Col>
          <Col className="col-8 text-right">
            <SubtaskFilterContainer />
          </Col>
        </Row>
        <SubtaskProgressBar subtaskList={subtaskList} />
        <SubtaskList
          taskId={taskId}
          taskStatus={taskStatus}
          subtaskList={subtaskList}
          currentFilter={currentFilter}
        />
        <SubtaskAdd taskId={taskId} taskStatus={taskStatus} />
      </Container>
    );
  }
}

const mapStateToProps = (
  {
    subtaskReducer: { subtaskList, currentFilter },
    taskDetailsReducer: { taskDetails: { _id: taskId, status: taskStatus } },
  },
) => ({
  subtaskList, currentFilter, taskId, taskStatus,
});

export { SubtaskListContainer as SubtaskListContainerComponent };
export default connect(
  mapStateToProps,
  null,
)(SubtaskListContainer);
