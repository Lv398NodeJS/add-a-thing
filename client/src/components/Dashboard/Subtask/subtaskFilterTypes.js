const SHOW_ALL = 'SHOW_ALL';
const SHOW_ACTIVE = 'SHOW_ACTIVE';
const SHOW_COMPLETED = 'SHOW_COMPLETED';

export {
  SHOW_ALL,
  SHOW_ACTIVE,
  SHOW_COMPLETED,
};
