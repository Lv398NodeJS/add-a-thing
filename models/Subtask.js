const mongoose = require('mongoose');

const { Schema } = mongoose;
const SubtaskSchema = new Schema({
  name: {
    type: String,
  },
  completed: Boolean,
  taskId: {
    type: Schema.Types.ObjectId,
    ref: 'Task',
    required: true,
  },
  date: {
    type: Date,
    default: Date.now,
  },
});

module.exports = Subtask = mongoose.model('Subtask', SubtaskSchema);
