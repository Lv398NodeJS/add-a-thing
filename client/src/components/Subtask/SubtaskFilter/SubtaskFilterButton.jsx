import React, { Component } from 'react';
import { Button } from 'react-bootstrap';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as subtaskActions from '@actions/subtaskActions';
import { SHOW_ACTIVE, SHOW_COMPLETED } from '../subtaskFilterTypes';

class SubtaskFilterButton extends Component {
  getTitle = () => {
    const { filterType } = this.props;

    switch (filterType) {
      case SHOW_ACTIVE:
        return 'Active';
      case SHOW_COMPLETED:
        return 'Completed';
      default:
        return 'All';
    }
  };

  isActive = () => {
    const { filterType, currentFilter } = this.props;
    return filterType === currentFilter;
  }

  render() {
    const { filterType, subtaskActions: { setSubtaskFilter } } = this.props;
    return (
      <Button size="sm" onClick={() => setSubtaskFilter(filterType)} active={this.isActive()}>{this.getTitle()}</Button>
    );
  }
}
const mapStateToProps = ({ subtaskReducer: { currentFilter } }) => ({
  currentFilter,
});
const mapDispatchToProps = dispatch => ({
  subtaskActions: bindActionCreators(subtaskActions, dispatch),
});

export { SubtaskFilterButton as SubtaskFilterButtonComponent };
export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SubtaskFilterButton);
