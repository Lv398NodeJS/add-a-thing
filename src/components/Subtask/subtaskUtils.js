import { SHOW_ALL, SHOW_ACTIVE, SHOW_COMPLETED } from './subtaskFilterTypes';

export const getSubtaskListAsArray = (snapValue = {}) => {
  if (typeof snapValue !== 'object' && snapValue.constructor !== Object) return [];
  const subtaskList = [];
  Object.keys(snapValue).forEach((subtask) => {
    const { text, completed } = snapValue[subtask];
    subtaskList.push({
      text,
      completed,
      id: subtask,
    });
  });
  return subtaskList;
};

export const getVisibleSubtasks = (subtaskList, currentFilter) => {
  switch (currentFilter) {
    case SHOW_ALL:
      return subtaskList;
    case SHOW_COMPLETED:
      return subtaskList.filter(subtask => subtask.completed);
    case SHOW_ACTIVE:
      return subtaskList.filter(subtask => !subtask.completed);
    default:
      throw new Error(`Unknown filter: ${currentFilter}`);
  }
}