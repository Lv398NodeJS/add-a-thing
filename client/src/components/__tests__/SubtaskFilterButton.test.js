import React from 'react';
import { shallow } from 'enzyme';
import { SubtaskFilterButtonComponent } from '../Dashboard/Subtask/SubtaskFilter/SubtaskFilterButton';
import { SHOW_ALL, SHOW_ACTIVE, SHOW_COMPLETED } from '../Dashboard/Subtask/subtaskFilterTypes';

describe('SubtaskFilterButton component', () => {
  const subtaskActions = { setSubtaskFilter: jest.fn() };
  it('should render correctly and match the snapshot', () => {
    const subtaskFilterButtonComponent = shallow(
      <SubtaskFilterButtonComponent subtaskActions={subtaskActions} />,
    );
    expect(subtaskFilterButtonComponent).toMatchSnapshot();
  });

  it('should render Bootstrap 4 Button', () => {
    const subtaskFilterButtonComponent = shallow(
      <SubtaskFilterButtonComponent subtaskActions={subtaskActions} />,
    );
    expect(subtaskFilterButtonComponent.find('Button').length).toBe(1);
  });

  it('should render Bootstrap 4 Button with text depending on "filterType" prop', () => {
    const subtaskFilterButtonComponent = shallow(
      <SubtaskFilterButtonComponent
        subtaskActions={subtaskActions}
        filterType={SHOW_ALL}
      />,
    );
    expect(subtaskFilterButtonComponent.find('Button').text()).toBe('All');
  });

  it('should render Bootstrap 4 Button with text depending on "filterType" prop', () => {
    const subtaskFilterButtonComponent = shallow(
      <SubtaskFilterButtonComponent
        subtaskActions={subtaskActions}
        filterType={SHOW_ACTIVE}
      />,
    );
    expect(subtaskFilterButtonComponent.find('Button').text()).toBe('Active');
  });

  it('should render Bootstrap 4 Button with text depending on "filterType" prop', () => {
    const subtaskFilterButtonComponent = shallow(
      <SubtaskFilterButtonComponent
        subtaskActions={subtaskActions}
        filterType={SHOW_COMPLETED}
      />,
    );
    expect(subtaskFilterButtonComponent.find('Button').text()).toBe('Completed');
  });

  it('should render active Bootstrap 4 Button if prop "filterType" is equal to prop "currentFilter"', () => {
    const subtaskFilterButtonComponent = shallow(
      <SubtaskFilterButtonComponent
        subtaskActions={subtaskActions}
        filterType={SHOW_COMPLETED}
        currentFilter={SHOW_COMPLETED}
      />,
    );
    expect(subtaskFilterButtonComponent.find('Button').prop('active')).toBeTruthy();
  });
});
