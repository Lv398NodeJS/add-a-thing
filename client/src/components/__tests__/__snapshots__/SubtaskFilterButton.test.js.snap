// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubtaskFilterButton component should render correctly and match the snapshot 1`] = `ShallowWrapper {}`;
